### PR TITLE
Support for Commodore 64 CRT and PRG files

### PR DESCRIPTION
**Pull request by mlund**
This adds support for program (PRG) and cartridge files (CRT) for the Commodore 64 8-bit computer.

### CODE REVIEW

**Comment by mmalecot** [comment]
Thanks for your PR! I will review it ASAP.

Before going into detail, I think the most appropriate `Kind` for PRG is `Executable`, don't you?

Also, for the CRT signature, you could perhaps enter it in ASCII since it is "C64 CARTRIDGE".

Finally, where did you find the PRG signature? Could there be others? I'll have a look at how the `file` command works (see file/magic/Magdir/c64).

**Comment by mlund** [comment]
Thanks for pointing to `magdir/c64`, I was actually looking for how to detect disk images. When loading on C64, the first two bytes gives the load address and for programs with a BASIC stub (i.e. `SYS ...`), this is 0x0801. It's not a very stringent test, and I see that `file` tries much harder. Yes, executable might be a better kind.

File: src/signatures.rs [hunk 1]
```diff
@@ -1371,6 +1374,9 @@ signatures! {
     format = BdavMpeg2TransportStream
     value = b"\x47" offset = 4, b"\x47" offset = 196
 
+    format = Commodore64Program
+    value = b"\x01\x08"
```
Comment (mmalecot) [inline src/signatures.rs:1378]: It seems that byte at offset 6 is always \x9E (this is also checked by the file command)

File: src/formats.rs [hunk 2]
```diff
@@ -515,6 +515,20 @@ formats! {
     extension = "clj"
     kind = Text
 
+    format = Commodore64Cartridge
+    name = "Commodore 64 cartridge file"
+    short_name = "CRT"
+    media_type = "application/x-commmodore-64-cartridge"
+    extension = "crt"
+    kind = Executable
```
Comment (mmalecot) [inline src/formats.rs:523]: Inverted kind between CRT and PRG

File: tests/rom.rs [hunk 3]
```diff
@@ -6,6 +6,18 @@ fn test_atari7800_rom() {
...
 
+#[test]
+fn test_commmodore64_cartridge() {
+    let fmt = FileFormat::from_file("fixtures/rom/sample.crt").unwrap();
+    assert_eq!(fmt, FileFormat::Commodore64Cartridge);
+}
+
+#[test]
+fn test_commodore_program() {
+    let fmt = FileFormat::from_file("fixtures/rom/sample.prg").unwrap();
```
Comment (mmalecot) [inline tests/rom.rs:17]: Fixture needs to be moved in fixtures/executable

**Review by mmalecot — COMMENTED** [review]
Review done :).

File: fixtures/rom/sample.prg
[the remark this answers is not in the document]
Comment (mlund) [reply fixtures/rom/sample.prg:1]: Yes, it's created by me ;-)

File: src/formats.rs [hunk 4]
```diff
@@ -515,6 +515,20 @@ formats! {
...
+
+    format = Commodore64Program
+    name = "Commodore 64 program file"
+    short_name = "PRG"
+    media_type = "application/x-commmodore-64-program"
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/formats.rs:528]: fixed

File: src/formats.rs [hunk 5]
```diff
@@ -515,6 +515,20 @@ formats! {
     extension = "clj"
     kind = Text
 
+    format = Commodore64Cartridge
+    name = "Commodore 64 cartridge file"
+    short_name = "CRT"
+    media_type = "application/x-commmodore-64-cartridge"
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/formats.rs:521]: fixed

File: src/formats.rs [hunk 6]
```diff
@@ -515,6 +515,20 @@ formats! {
     extension = "clj"
     kind = Text
 
+    format = Commodore64Cartridge
+    name = "Commodore 64 cartridge file"
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/formats.rs:519]: fixed

File: src/formats.rs [hunk 7]
```diff
@@ -515,6 +515,20 @@ formats! {
...
+    short_name = "CRT"
+    media_type = "application/x-commmodore-64-cartridge"
+    extension = "crt"
+    kind = Executable
+
+    format = Commodore64Program
+    name = "Commodore 64 program file"
```
[the remark this answers is not in the document]
Comment (mlund) [reply src/formats.rs:526]: fixed

File: tests/rom.rs [hunk 8]
```diff
@@ -6,6 +6,18 @@ fn test_atari7800_rom() {
...
 }
 
+#[test]
+fn test_commmodore64_cartridge() {
+    let fmt = FileFormat::from_file("fixtures/rom/sample.crt").unwrap();
+    assert_eq!(fmt, FileFormat::Commodore64Cartridge);
+}
+
+#[test]
+fn test_commodore_program() {
```
[the remark this answers is not in the document]
Comment (mlund) [reply tests/rom.rs:16]: fixed and moved to executable.rs

File: src/signatures.rs (hunk 1 above)
Comment (mlund) [reply src/signatures.rs:1378]: Updated and moved signature to 3-byte section

File: src/formats.rs (hunk 2 above)
Comment (mlund) [reply src/formats.rs:523]: fixed

File: tests/rom.rs (hunk 3 above)
Comment (mlund) [reply tests/rom.rs:17]: fixed

File: fixtures/rom/sample.crt
[the remark this answers is not in the document]
Comment (mlund) [reply fixtures/rom/sample.crt:1]: It's the first 50 bytes of a 260 KB CRT file containing a custom version of a game, probably pirated back in the 1980s. I'm pretty confident there's no copyrighted code, but we could truncate even further if you prefer(?).

File: src/signatures.rs [hunk 9]
```diff
@@ -1277,6 +1280,10 @@ signatures! {
     format = Bzip2
     value = b"BZh"
 
+    format = Commodore64Program
+    value = b"\x01\x08"
```
Comment (mmalecot) [inline src/signatures.rs:1284]: Several `value =` means "or".

The syntax for an "and" is done using comma, e.g.: `value = b"\x01\x08", b"\x9E" offset = 6`.

**Review by mmalecot — COMMENTED** [review]
Almost good!

File: fixtures/rom/sample.crt
[the remark this answers is not in the document]
Comment (mmalecot) [reply fixtures/rom/sample.crt:1]: It's up to you

[the remark this answers is not in the document]
Comment (mlund) [reply fixtures/rom/sample.crt:1]: Let's keep it as we are only including part of the header, nothing from the game.

**Pull request merged**